homeassistant/components/google/__init__.py

```diff
@@ -159,131 +147,79 @@ def scope(self) -> str:
 )
 
 
-def do_authentication(
-    hass: HomeAssistant,
-    hass_config: ConfigType,
-    config: ConfigType,
-    storage: Storage,
-) -> bool:
-    """Notify user of actions and authenticate.
-
-    Notify user of user_code and verification_url then poll
-    until we have an access token.
-    """
-    oauth = OAuth2WebServerFlow(
-        client_id=config[CONF_CLIENT_ID],
-        client_secret=config[CONF_CLIENT_SECRET],
-        scope=config[CONF_CALENDAR_ACCESS].scope,
-        redirect_uri="Home-Assistant.io",
-    )
-    try:
-        dev_flow = oauth.step1_get_device_and_user_codes()
-    except OAuth2DeviceCodeError as err:
-        persistent_notification.create(
-            hass,
-            f"Error: {err}<br />You will need to restart hass after fixing." "",
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID,
-        )
-        return False
-
-    persistent_notification.create(
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Google component."""
+    conf = config.get(DOMAIN, {})
+    hass.data[DOMAIN] = {DATA_CONFIG: conf}
+    config_flow.OAuth2FlowHandler.async_register_implementation(
         hass,
-        (
-            f"In order to authorize Home-Assistant to view your calendars "
-            f'you must visit: <a href="{dev_flow.verification_url}" target="_blank">{dev_flow.verification_url}</a> and enter '
-            f"code: {dev_flow.user_code}"
+        DeviceAuth(
+            hass,
+            conf[CONF_CLIENT_ID],
+            conf[CONF_CLIENT_SECRET],
         ),
-        title=NOTIFICATION_TITLE,
-        notification_id=NOTIFICATION_ID,
     )
 
-    listener: CALLBACK_TYPE | None = None
-
-    def step2_exchange(now: datetime) -> None:
-        """Keep trying to validate the user_code until it expires."""
-        _LOGGER.debug("Attempting to validate user code")
-
-        # For some reason, oauth.step1_get_device_and_user_codes() returns a datetime
-        # object without tzinfo. For the comparison below to work, it needs one.
-        user_code_expiry = dev_flow.user_code_expiry.replace(tzinfo=timezone.utc)
-
-        if now >= user_code_expiry:
-            persistent_notification.create(
-                hass,
-                "Authentication code expired, please restart "
-                "Home-Assistant and try again",
-                title=NOTIFICATION_TITLE,
-                notification_id=NOTIFICATION_ID,
+    # Import credentials from the old token file into the new way as
+    # a ConfigEntry managed by home assistant.
+    storage = Storage(hass.config.path(TOKEN_FILE))
+    creds = await hass.async_add_executor_job(storage.get)
+    if creds and conf[CONF_CALENDAR_ACCESS].scope in creds.scopes:
+        _LOGGER.debug("Importing configuration entry with credentials")
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
```

Comment:
@MartinHjelmare I imagine i should delete the TOKEN_FILE after import? My impression is the import can just happen once, then it is persisted, so it would obsolete the token file once that is done.  Perhaps in a follow up PR I can handle this.

Reply:
I don't think we should delete it to allow users to roll back. Since we guard for single config entry there will only be one import completed. If we like we can check for existing config entries before starting the import.